File: Json/src/main/java/org/elkoserver/json/JsonParsing.java
package org.elkoserver.json;

import com.grack.nanojson.JsonParser;
import com.grack.nanojson.JsonParserException;

import java.io.IOException;
import java.io.Reader;

public class JsonParsing {
    public static JsonObject jsonObjectFromString(String string) throws JsonParserException {
        JsonObject result;

        if (string.isEmpty()) {
            result = null;
        } else {
            com.grack.nanojson.JsonObject nanoObject = JsonParser.object().from(string);
            result = nanoObject == null ? null : new JsonObject(nanoObject);
        }

        return result;
    }

    public static JsonObject jsonObjectFromReader(Reader reader) throws JsonParserException {
        JsonObject result;

        if (!reader.markSupported()) {
            throw new IllegalStateException();
        }

        int nextCharacter;
        try {
            reader.mark(1);
            nextCharacter = reader.read();
            reader.reset();
        } catch (IOException e) {
            throw new IllegalArgumentException(e);
        }

        if (nextCharacter == -1) {
            result = null;
        } else {
            com.grack.nanojson.JsonObject nanoObject = JsonParser.object().from(reader);
            result = nanoObject == null ? null : new JsonObject(nanoObject);
        }

        return result;
    }
}

File: Json/src/main/kotlin/org/elkoserver/json/JsonParsing.kt
package org.elkoserver.json

import com.grack.nanojson.JsonObject
import com.grack.nanojson.JsonParser
import com.grack.nanojson.JsonParserException
import java.io.IOException
import java.io.Reader

object JsonParsing {
    @Throws(JsonParserException::class)
    fun jsonObjectFromString(string: String) =
            if (string.isEmpty()) {
                null
            } else {
                JsonParser.`object`().from(string)
            }

    @Throws(JsonParserException::class)
    fun jsonObjectFromReader(reader: Reader): JsonObject? {
        val result: JsonObject?
        check(reader.markSupported())
        val nextCharacter: Int
        try {
            reader.mark(1)
            nextCharacter = reader.read()
            reader.reset()
        } catch (e: IOException) {
            throw IllegalArgumentException(e)
        }
        result = if (nextCharacter == -1) {
            null
        } else {
            val nanoObject = JsonParser.`object`().from(reader)
            nanoObject?.let(::JsonObject)
        }
        return result
    }
}
